### PR TITLE
[WIP] Bug 1215568: Implement an in-memory Azure table for testing

### PR DESCRIPTION
**Pull request by djmitche**
@jonasfj, if you don't mind taking a look at this very rough work-in-progress, I'd appreciate it.  I'm doing this without reference to the Azure docs, just for fun.  So far, the create, load, and modify tests all pass against both Azure and the in memory fake.

I'm a little concerned about the eTag handling though, and in particular the `_.omit` hack.  Do you have a suggestion of a better, but still simple, way to handle that?

Also, I realize that so far I'm just writing a fake fast-azure-storage.  If that continues to be the case, I'll refactor this to either be a part of FAS or another package.  But when I implement filtering, I think I will find it a lot easier to fake things out before the filter expression is turned into a string.

### CODE REVIEW

**Comment by jonasfj** [comment]
if this ended up being a fake `fast-azure-storage` we should totally add that to ` fast-azure-storage` :)

It would be quiet a killer feature...

**Comment by djmitche** [comment]
Cool -- I'll see if I can parse the filter expressions then.

File: src/inmemory.js [hunk 1]
```diff
@@ -0,0 +1,224 @@
...
+
+/**
+ * Create table.
+ *
+ * @method createTable
+ * @return {Promise} A promise that the table was created.
+ */
+InMemoryWrapper.prototype.createTable = function() {
+  tables[this.table] = {};
+  // TODO: EntityAlready Exists?
```
Comment (jonasfj) [inline src/inmemory.js]: yes, global tables are bad.. if they we always overwrite it :)

So not doing `tables[this.table] = {};` if `tables[this.table] != undefined` might be smart..
Returning the exists error is also good... just for making tests more accurate, I think we do handle the error correctly.

File: src/inmemory.js [hunk 2]
```diff
@@ -0,0 +1,224 @@
...
+};
+
+/**
+ * Delete table.
+ *
+ * @method deleteTable
+ * @return {Promise} A promise that the table was marked for deletion.
+ */
+InMemoryWrapper.prototype.deleteTable = function() {
+  // TODO: 404?
```
Comment (jonasfj) [inline src/inmemory.js]: Again, yes... we do handle the 404 error... IMO it's stupid of azure to have it, because delete should be idempotent... but apparently they like footguns. So doing 404 would increase our test coverage.

File: src/inmemory.js [hunk 3]
```diff
@@ -0,0 +1,224 @@
...
+      _.forIn(entity, function(val, prop) {
+        existing[prop] = val;
+      });
+      entity = existing;
+    }
+  } else {
+    if (!options.eTag) {
+      tables[this.table][key] = entity;
+    } else if (options.eTag != '*') {
+      return Promise.reject(makeError(500, 'NotSureWhatToDo'));
```
Comment (jonasfj) [inline src/inmemory.js]: 404 entity doesn't exist or maybe it's a 412, see fast-azure-storage tests, or add a test case to it... to make sure what it is...

File: test/create_load_test.js [hunk 4]
```diff
@@ -7,35 +7,13 @@ var Promise = require('promise');
...
-  });
-
-  var Item;
-  test("Item = ItemV1.setup", function() {
-    Item = ItemV1.setup({
-      credentials:  helper.cfg.azure,
-      table:        helper.cfg.tableName
-    });
-  });
+var makeTests = function(Item, Item2, drain) {
```
Comment (jonasfj) [inline test/create_load_test.js]: I wish this could be prettier... but the concept it great

File: src/inmemory.js [hunk 5]
```diff
@@ -0,0 +1,224 @@
...
+};
+
+/* Internal utilities */
+
+var entityEtag = function(entity) {
+  var sha1 = crypto.createHash('sha1');
+  // filter out metadata
+  var odata = /^odata\.|@odata/;
+  entity = _.omit(entity, function(v, k) { return odata.test(k); });
+  sha1.update(JSON.stringify(entity));
```
Comment (jonasfj) [inline src/inmemory.js]: Use `json-stable-stringify`, using `JSON.stringify()` will bite you badly one day :)

**Comment by jonasfj** [comment]
@djmitche, this generally looks like a sane approach and you're already very far :)

I see the etag issue.. The sanest thing would be to do defaults for the `<property>@odata.type` fields...
and always have an `<property>@odata.type` field... Really, we're only required to add the `<property>@odata.type`  in some cases... but it's should be easy to infer the defaults in cases where it's left out, see:
https://msdn.microsoft.com/en-us/library/azure/dn535600.aspx

Basically, if `<property>` is a boolean it's `Edm.Boolean`, if it's a number with a decimal point it's `Edm.Double`, and if it's a number without a decimal point it's a `Edm.Int32`, if it's a string it's a `Edm.String`. That's it...

Note, I think it's sane to omit all the odata fields that doesn't concern custom properties...

File: src/inmemory.js (hunk 3 above)
Comment (djmitche) [inline src/inmemory.js]: That's my plan for all of the unknowns -- write a test case, assert what I see from Azure, and then make the fake pass the test :)

**Comment by djmitche** [comment]
Thanks -- good advice all!

File: src/inmemory.js (hunk 3 above)
Comment (jonasfj) [inline src/inmemory.js]: > That's my plan for all of the unknowns

Then maybe it makes sense to add this to `fast-azure-storage`... So we can have the same test cases for both in-memory and azure-backed...

File: test/create_load_test.js (hunk 4 above)
Comment (djmitche) [inline test/create_load_test.js]: I'll try to play with it a bit -- I could potentially have a helper that calls suite() twice

Comment (jonasfj) [inline test/create_load_test.js]: > I could potentially have a helper that calls suite() twice

Yes, that would be lovely...

Something like:
```js
MultiSuite('Entity (create/load)', [
  {
    case: "azure-credentials",
    options: {credentials: ..., table: ...},
  }, {
    case: "in-memory",
    options: {inmemory: true, table: ...},
  }
], (options) => {
  Item = ItemV1.configure(options);
  ...
});
```

Though `MultiSuite` is a really bad name... Maybe it's a helper we can add to `taskcluster-lib-testing`, if we can find a proper name :)

Comment (djmitche) [inline test/create_load_test.js]: Here's what I came up with:
```
helper.contextualSuites("Entity (modify)", [
  {
    context: "Azure",
    options: function() {
      return {
        Item: Item.setup({
          credentials:  helper.cfg.azure,
          table:        helper.cfg.tableName
        })  
      };  
    }   
  }, {
    context: "Azure",
    options: function() {
      return {
        Item: Item.setup({
          inMemory: true,
          table:    'items'
        })  
      };  
    }   
  }
], function(context, options) {
  var Item = options.Item;

  setup(function() {
    return Item.ensureTable();
  }); 

  test("Item.create, Item.modify, Item.load", function() {
    var id = slugid.v4();
```

It basically calls the options function, then calls the suite function with its result..

**Comment by djmitche** [comment]
Lots more tests now, but mostly around data types.  Still lots of TODO!

I think I'm going to implement this within azure-entities, since fast-azure-storage is quite a bit more general and I don't want to have to implement the entire odata standard (including parsing filters and implementing all those functions, ugh).  The bits that azure-entities uses are much more limited.

**Comment by djmitche** [comment]
As always, ignore the TC results.  Right now:
```
  353 passing (40s)
  16 pending
```
and those 16 pending are because I haven't implemented the query operation yet.  I'm going to do that with an un-parsed filter, and just implement the filtering supported by azure-entities.  Basically, enough to support the test cases here and anything I encounter testing tc-hooks and tc-secrets, and no more.

**Comment by djmitche** [comment]
```
  369 passing (38s)
```

and ready for review.

As discussed in irc, I've only implemented things in the fake that are actually used by azure-entities.  I don't want to go substantially beyond that, and I'm willing to accept some risk of divergent behavior between the two, which we can fix as we find it.

I'm open to suggestions of alternative implementation strategies.  I'd be really excited for suggestions of test cases that I should add which might show important differences in behavior between the fake and Azure.

File: src/entity.js [hunk 6]
```diff
@@ -570,6 +574,17 @@ Entity.configure = function(options) {
  *   },
  * }
  *
+ * To use an in-memory, testing-oriented table, replace the `account` key with
+ * `inMemory`, and do not supply credentials:
+ * {
+ *   inMemory: true,
```
Comment (jonasfj) [inline src/entity.js]: Suggestion instead of specifying `inMemory: true` we could use `account: 'in-memory'` so it becomes of a special case of specifying `account`, `table` and taskcluster `credentials`.

This way our config files become a lot simpler... As we just specify `account: 'in-memory'` and in production `account: 'taskclusterqueue'`...

File: src/entity.js [hunk 7]
```diff
@@ -677,101 +693,113 @@ Entity.setup = function(options) {
...
-          process:      options.process,
-          duration:     d[0] * 1000 + (d[1] / 1000000),
-          table:        options.table,
-          method:       name,
-          error:        (err ? err.code : null) || 'UnknownError'
-        });
-        throw err;
+  if (options.inMemory) {
+    if (!inmemory) {
+      inmemory = require('./inmemory'); // lazy-loaded
```
Comment (jonasfj) [inline src/entity.js]: There is no need to make this a global reference... `require` won't reload, and it's super expensive...
Would be worth it if we were referencing `inmemory` elsewhere, but we aren't...

File: src/entity.js [hunk 8]
```diff
@@ -677,101 +693,113 @@ Entity.setup = function(options) {
...
+    }
+    subClass.prototype.__table = options.table;
+    subClass.prototype.__filterBuilder = inmemory.appendFilter;
+    subClass.prototype.__aux = new inmemory.InMemoryWrapper(options.table);
+  } else {
```
Comment (jonasfj) [inline src/entity.js]: Since this else extends to the end of the function can we please do an early return instead... and avoid the extra level of indentation...

File: src/inmemory.js [hunk 9]
```diff
@@ -0,0 +1,367 @@
...
+      return;
+    }
+    var odata_type = k + '@odata.type';
+    if (entity[odata_type]) {
+      return;
+    }
+    if (typeof v === "boolean") {
+      entity[odata_type] = 'Edm.Boolean';
+    } else if (typeof v === "number") {
+      if (/\./.test(v)) {
```
Comment (jonasfj) [inline src/inmemory.js]: I think it's a bit saner to do `v % 1 === 0`

File: src/inmemory.js [hunk 10]
```diff
@@ -0,0 +1,367 @@
...
+ */
+InMemoryWrapper.prototype.insertEntity = function(entity) {
+  var key = makeKey(entity.PartitionKey, entity.RowKey);
+  if (!tables[this.table]) {
+    return Promise.reject(makeError(404, 'ResourceNotFound'));
+  }
+  if (key in tables[this.table]) {
+    return Promise.reject(makeError(409, 'EntityAlreadyExists'));
+  }
+  updateTimestamp(entity);
```
Comment (jonasfj) [inline src/inmemory.js]: Probably best to do an `entity = _.cloneDeep(entity)` before storing it... Probably the same in modify...

File: src/inmemory.js [hunk 11]
```diff
@@ -0,0 +1,367 @@
...
+ * }
+ * ```
+ * @return {Promise} A promise for `eTag` of the modified entity.
+ */
+InMemoryWrapper.prototype.updateEntity = function(entity, options) {
+  var key = makeKey(entity.PartitionKey, entity.RowKey);
+  if (!tables[this.table]) {
+    return Promise.reject(makeError(404, 'ResourceNotFound'));
+  }
+  if (key in tables[this.table]) {
```
Comment (jonasfj) [inline src/inmemory.js]: Generally safer to do: `if(tables[this.table][key]) {`

Just in case you set the key null rather than delete it...

**Comment by jonasfj** [comment]
See comments mostly this look good... I think my comments are pure nits...

File: src/inmemory.js (hunk 10 above)
Comment (jonasfj) [inline src/inmemory.js]: Actually not sure this is needed... but it really can't hurt just to be safe :)

**Comment by jonasfj** [comment]
btw, I think a solid test it to run the queue with this...

File: src/entity.js (hunk 6 above)
Comment (djmitche) [inline src/entity.js]: I had been expecting to not use a configuration file for tests, but now that you mention it it might be nice if we can run the same tests in both the real-azure and fake.

File: src/entity.js (hunk 7 above)
Comment (djmitche) [inline src/entity.js]: I'm trying to avoid loading inmemory at all unless it's used (so, not loaded in production).  Does it make sense in that case?

File: src/inmemory.js (hunk 10 above)
Comment (djmitche) [inline src/inmemory.js]: They get cloned when returned.. but yeah, can't hurt.

File: src/inmemory.js (hunk 3 above)
Comment (djmitche) [inline src/inmemory.js]: Heh, this one needs a test case

**Comment by djmitche** [comment]
OK, so

* [ ] allow to specify inMemory from the config file
* [ ] early return in `setup`
* [ ] `v % 1 === 0` to test for decimal
* [ ] `entity = _.cloneDeep(entity)` on insert/update
* [ ] `if(tables[this.table][key])` rather than `key in ..`
* [ ] test case for `NotSureWhatToDo`
* [ ] test with secrets
* [ ] test with queue

**Comment by djmitche** [comment]
It looks like the queue uses fast-azure-storage directly for at least some of its operations, so it's not a suitable test case here.  However, the taskcluster-hooks test suite ran just fine against this branch -- and in the process made 0 outbound network connections, meaning anyone can run the tests!

Given that, I think we should get this merged and shipped to npm.

**Comment by jonasfj** [comment]
Yeah, ship it :)
Den 20. jan. 2016 7.28 PM skrev "Dustin J. Mitchell" <
notifications@github.com>:
